File: src/main/java/sk/hudak/prco/dao/db/impl/GroupOfProductFindEntityDaoImpl.java
package sk.hudak.prco.dao.db.impl;

import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import sk.hudak.prco.dao.db.GroupOfProductFindEntityDao;
import sk.hudak.prco.model.ProductEntity;
import sk.hudak.prco.model.QGroupEntity;
import sk.hudak.prco.model.QGroupOfProductFindEntity;
import sk.hudak.prco.model.QProductEntity;

import javax.persistence.EntityManager;
import java.util.List;
import java.util.Optional;

@Component
public class GroupOfProductFindEntityDaoImpl implements GroupOfProductFindEntityDao {

    @Autowired
    private EntityManager em;

    @Override
    public List<ProductEntity> findProductsWitchAreNotInAnyGroup() {
        JPAQueryFactory queryFactory = getQueryFactory();
        return queryFactory
                .select(QProductEntity.productEntity)
                .from(QProductEntity.productEntity)
                .where(QProductEntity.productEntity.id.notIn(getCountOfProductWhichAreInAtLeastOneGroup(queryFactory)))
                .fetch();
    }

    @Override
    public long countOfProductsWitchAreNotInAnyGroup() {
        JPAQueryFactory queryFactory = getQueryFactory();
        return queryFactory
                .select(QProductEntity.productEntity)
                .from(QProductEntity.productEntity)
                .where(QProductEntity.productEntity.id.notIn(getCountOfProductWhichAreInAtLeastOneGroup(queryFactory)))
                .fetchCount();
    }

    private List<Long> getCountOfProductWhichAreInAtLeastOneGroup(JPAQueryFactory queryFactory) {
        return queryFactory
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.productId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .distinct()
                .fetch();
    }

    @Override
    public long countOfProductInGroup(String groupName) {
        return getQueryFactory()
                .select(QGroupEntity.groupEntity)
                .from(QGroupEntity.groupEntity)
                .where(QGroupEntity.groupEntity.name.eq(groupName))
                .fetchFirst().getProducts().size();
    }

    @Override
    public Optional<Long> findFirstProductGroupId(Long productId) {
        return Optional.ofNullable(getQueryFactory()
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.groupId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .where(QGroupOfProductFindEntity.groupOfProductFindEntity.productId.eq(productId))
                .fetchFirst());
    }

    @Override
    public List<Long> findGroupIdsWithProductId(Long productId) {
        return getQueryFactory()
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.groupId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .where(QGroupOfProductFindEntity.groupOfProductFindEntity.productId.eq(productId))
                .fetch();
    }

    private JPAQueryFactory getQueryFactory() {
        return new JPAQueryFactory(em);
    }

}

File: src/main/java/sk/hudak/prco/dao/db/impl/GroupOfProductFindEntityDaoImpl.kt
package sk.hudak.prco.dao.db.impl

import com.querydsl.core.types.Order
import com.querydsl.core.types.OrderSpecifier
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import sk.hudak.prco.dao.db.GroupOfProductFindEntityDao
import sk.hudak.prco.model.ProductEntity
import sk.hudak.prco.model.QGroupEntity
import sk.hudak.prco.model.QGroupOfProductFindEntity
import sk.hudak.prco.model.QProductEntity
import sk.hudak.prco.utils.DateUtils
import java.util.*
import javax.persistence.EntityManager

@Repository
open class GroupOfProductFindEntityDaoImpl(private val em: EntityManager) : GroupOfProductFindEntityDao {

    private fun getQueryFactory() = JPAQueryFactory(em)

    override fun findProductsWitchAreNotInAnyGroup(applyExpiration: Boolean): List<ProductEntity> {
        val queryFactory = getQueryFactory()
        return queryFactory
                .select(QProductEntity.productEntity)
                .from(QProductEntity.productEntity)
                .where(QProductEntity.productEntity.id.notIn(
                        queryFactory
                                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.productId)
                                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                                .distinct()
                ).and(QProductEntity.productEntity.lastTimeDataUpdated.isNull
                        .or(QProductEntity.productEntity.lastTimeDataUpdated.lt(DateUtils.calculateDate(12))))
                )
                .orderBy(OrderSpecifier(Order.DESC, QProductEntity.productEntity.created))
                .fetch()
    }

    override fun findProductsWitchAreNotInAnyGroup(): List<ProductEntity> {
        val queryFactory = getQueryFactory()
        return queryFactory
                .select(QProductEntity.productEntity)
                .from(QProductEntity.productEntity)
                .where(QProductEntity.productEntity.id.notIn(findIdsOfProductWhichAreInAtLeastOneGroup(queryFactory)))
                .orderBy(OrderSpecifier(Order.DESC, QProductEntity.productEntity.created))
                .fetch()
    }

    override fun countOfProductsWitchAreNotInAnyGroup(): Long {
        val queryFactory = getQueryFactory()
        return queryFactory
                .select(QProductEntity.productEntity)
                .from(QProductEntity.productEntity)
                .where(QProductEntity.productEntity.id.notIn(findIdsOfProductWhichAreInAtLeastOneGroup(queryFactory)))
                .fetchCount()
    }

    private fun findIdsOfProductWhichAreInAtLeastOneGroup(queryFactory: JPAQueryFactory): List<Long> {
        return queryFactory
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.productId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .distinct()
                .fetch()
    }

    override fun countOfProductInGroup(groupName: String): Long {
        return getQueryFactory()
                .select(QGroupEntity.groupEntity)
                .from(QGroupEntity.groupEntity)
                .where(QGroupEntity.groupEntity.name.eq(groupName))
                .fetchFirst().products.size.toLong()
    }

    override fun findFirstProductGroupId(productId: Long?): Optional<Long> {
        return Optional.ofNullable(getQueryFactory()
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.groupId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .where(QGroupOfProductFindEntity.groupOfProductFindEntity.productId.eq(productId!!))
                .fetchFirst())
    }

    override fun findGroupIdsWithProductId(productId: Long): List<Long> {
        return getQueryFactory()
                .select(QGroupOfProductFindEntity.groupOfProductFindEntity.groupId)
                .from(QGroupOfProductFindEntity.groupOfProductFindEntity)
                .where(QGroupOfProductFindEntity.groupOfProductFindEntity.productId.eq(productId))
                .fetch()
    }

}
